### PR TITLE
feat(v2): allow non sidebar category to be first item of sidebar

### PR DESCRIPTION
**Pull request by endiliey**
## Motivation

This allows user to define sidebar item as the first level item of a sidebar. If previously its defined like this
```typescript
type Sidebar = {
  [sidebarId: string]: {
    [sidebarCategory: string]: SidebarItem[];
  }
}
```

Its now
```typescript
type Sidebar = {
  [sidebarId: string]: {
    [sidebarCategory: string]: SidebarItem[];
  } | SidebarItem[];
}
```

Other changes
- I added a lot of other tests in sidebars. So now if there is any wrong type of 'label', 'items', unknown type and so on, user will receive MUCH BETTER error messages. This is still 100% test coverage on docs
 Updated babel so i can use TS 3.7 assertion
- Docs updated

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

all test pass

```js
// sidebars.js
module.exports = {
    docs: [
    {
      type: 'category',
      label: 'Docusaurus',
      items: ['introduction', 'design-principles', 'contributing'],
    },
    {
      type: 'category',
      label: 'Getting Started',
      items: ['installation', 'configuration'],
    },
    'creating-pages',
    'cli',
     {
        type: 'category',
        label: 'Plugins',
        items: ['using-plugins', 'advanced-plugins'],
      },
  ],
};
```

first item not category

Here's a cool demo showing it. You can directly click "Creating Pages"
**gif sidebar item**

### CODE REVIEW

File: website/docs/sidebar.md [hunk 1]
```diff
@@ -26,7 +26,18 @@ module.exports = {
...
+
+```typescript
+type Sidebar = {
+  [sidebarId: string]: {
+    [sidebarCategory: string]: SidebarItem[];
+  } | SidebarItem[];
+}
+```
+
+Below is an example of a sidebar object. The key `docs` is the id of the sidebar (can be renamed to something else) and `Getting Started` is a category within the sidebar. `greeting` and `doc1` is just a [Sidebar Item](#sidebar-item).
```
Comment (wgao19) [inline website/docs/sidebar.md]: grammar: greeting and doc1 -are- sidebar items.

(im on a mobile device so not convenient to commit directly)

Comment (endiliey) [inline website/docs/sidebar.md]:
```suggestion
Below is an example of a sidebar object. The key `docs` is the id of the sidebar (can be renamed to something else) and `Getting Started` is a category within the sidebar. `greeting` and `doc1` are both [sidebar item](#sidebar-item).
```

File: packages/docusaurus-theme-classic/src/theme/DocSidebar/index.js [hunk 2]
```diff
@@ -171,19 +171,16 @@ function DocSidebar(props) {
           )}
         </button>
         <ul className="menu__list">
-          {sidebarData.map(
-            item =>
-              item.items.length > 0 && (
```
Comment (yangshun) [inline packages/docusaurus-theme-classic/src/theme/DocSidebar/index.js]: Do we not need to check anymore?

Comment (endiliey) [reply packages/docusaurus-theme-classic/src/theme/DocSidebar/index.js]: if the first item is not a category, items is undefined. Checking is alrd done in the DocSidebarItem itself

Comment (endiliey) [reply packages/docusaurus-theme-classic/src/theme/DocSidebar/index.js]: without this loc removal, it will error

**Comment by yangshun** [comment]
Ok please resolve conflicts and self-merge :D

**Comment by endiliey** [comment]
Yeah that will have to wait maybe tonight. I have plans for today

**Comment by flaurida** [comment]
thank you! 🎉

**Pull request merged**